src/shell/js/composeui-fdc3/src/ComposeUIDesktopAgent.ts

```diff
@@ -0,0 +1,312 @@
+
+            const listener = <ComposeUIListener>await this.currentChannel!.addContextListener(stringContextType, handler!);
+            await this.currentChannel!.getCurrentContext(stringContextType)
+                .then(async (resultContext) => {
+                    listener.LatestContext = this.currentChannel!.retrieveCurrentContext(stringContextType);
+                    if (resultContext != listener.LatestContext) {
+                        //TODO: integrationtest
+                        await listener.handleContextMessage();
+                    } else {
+                        await listener.handleContextMessage(resultContext);
+                    }
+                });
+
+            this.currentChannelListeners.push(listener);
+            resolve(listener);
+        });
+    }
+
+    public getUserChannels(): Promise<Array<Channel>> {
+        return Promise.resolve(this.userChannels);
+    }
+
+    //TODO: should return AccessDenied error when a channel object is denied?
+    public joinUserChannel(channelId: string): Promise<void> {
+        return new Promise<void>(async (resolve, reject) => {
+            if (this.currentChannel) {
+                reject(new Error(ChannelError.AccessDenied));
+            }
+
+            let channel = this.userChannels.find(innerChannel => innerChannel.id == channelId);
+            if (!channel) {
+                try{
+                    await this.invokeChannelCreationMessage(ComposeUITopic.joinUserChannel(), channelId, "user");
+                    resolve();
+                } catch(error) {
+                    reject(error);
+                }
+            } else {
+                this.currentChannel = channel;
+                resolve();
+            }
+        });
+    }
+
+    //TODO: should return AccessDenied error when a channel object is denied
+    //TODO: should return a CreationFailed error when a channel cannot be created or retrieved (channelId failure)
+    public getOrCreateChannel(channelId: string): Promise<Channel> {
+        // return new Promise<Channel>(async(resolve, reject) => {
+        //     let channel = this.appChannels.find(innerChannel => innerChannel.id == channelId);
+        //     if (!channel) {
+        //         try{
+        //             await this.invokeChannelCreationMessage(ComposeUITopic.getOrCreateChannel(), channelId, "app");
+        //         } catch(error) {
+        //             reject(error);
+        //         }
+        //         channel = new ComposeUIChannel(channelId, "app", this.messageRouterClient);
+        //         this.addChannel(channel);
+        //     }
+        //     resolve(channel);
+        // });
+        throw new Error("Not implemented.");
+    }
+
+    //TODO
+    public createPrivateChannel(): Promise<PrivateChannel> {
+        throw new Error("Not implemented");
+    }
+
+    public getCurrentChannel(): Promise<Channel | null> {
+        return Promise.resolve(this.currentChannel!);
```

Comment:
I think I made you change the type of `currentChannel`, but here I can see I was wrong: it should be `currentChannel: Channel | null`. 
Alternatively, you can resolve with `this.currentChannel ?? null` to make sure that `undefined` is never returned.